virttest/libvirt_xml/vm_xml.py

```diff
@@ -613,41 +613,44 @@ def vm_rename(vm, new_name, uuid=None, virsh_instance=base.virsh):
         :param vm: VM class type instance
         :param new_name: new name of vm
         :param uuid: new_vm's uuid, if None libvirt will generate.
-        :return: a new VM instance
+        :return: a new VM instance or raise LibvirtXMLError
         """
+        if vm.name == new_name:
+            logging.warning("New name is same as VM, No need to rename")
+            return vm
+        start_vm = False
         if vm.is_alive():
             vm.destroy(gracefully=True)
-        vmxml = VMXML.new_from_dumpxml(vm_name=vm.name,
-                                       virsh_instance=virsh_instance)
+            start_vm = True
+        vmxml = VMXML.new_from_dumpxml(vm.name, virsh_instance=virsh_instance)
         backup = vmxml.copy()
-        # can't do in-place rename, must operate on XML
+
+        def _cleanup(details=""):
+            backup.define()
+            if start_vm:
+                vm.start()
+            raise xcepts.LibvirtXMLError("%s" % details)
+
+        # Undefine old VM firstly
         if not vmxml.undefine():
-            del vmxml  # clean up temporary files
-            raise xcepts.LibvirtXMLError("Error reported while undefining VM")
+            _cleanup(details="Undefine VM %s failed" % vm.name)
         # Alter the XML
         vmxml.vm_name = new_name
         if uuid is None:
-            # invalidate uuid so libvirt will regenerate
+            # UUID will be regenerated automatically
             del vmxml.uuid
-            vm.uuid = None
         else:
             vmxml.uuid = uuid
-            vm.uuid = uuid
-        # Re-define XML to libvirt
-        logging.debug("Rename %s to %s.", vm.name, new_name)
-        # error message for failed define
-        error_msg = "Error reported while defining VM:\n"
-        try:
-            if not vmxml.define():
-                raise xcepts.LibvirtXMLError(error_msg + "%s"
-                                             % vmxml.get('xml'))
-        except process.CmdError, detail:
-            del vmxml  # clean up temporary files
-            # Allow exceptions thrown here since state will be undefined
-            backup.define()
-            raise xcepts.LibvirtXMLError(error_msg + "%s" % detail)
-        # Keep names uniform
+        if not vmxml.define():
+            _cleanup(details="Define VM %s failed" % new_name)
+        # Update the name and uuid property for VM object
         vm.name = new_name
+        vm.uuid = VMXML.new_from_dumpxml(new_name,
+                                         virsh_instance=virsh_instance).uuid
+        if uuid is not None and vm.uuid != uuid:
+            _cleanup(details="UUID %s is not excepted %s" % (vm.uuid, uuid))
```

Comment:
Is it right to fail out if the UUID is not exactly same? libvirt supports a lot of different formats of UUID. with or without hyphen(`-`), lower or upper case. I think we can't check the validity just by string comparison.